chcemvediet/apps/inforequests/admin.py

```diff
@@ -384,21 +388,65 @@ def get_queryset(self, request):
     def get_inforequest(self, obj):
         return obj.branch.inforequest
 
+    def delete_warnings(self, obj):
+        warnings = []
+        if not obj.is_last_action:
+            warnings.append(format_html(squeeze(u"""
+                    {} is not the last action in the branch. Deleting it may cause logical errors in
+                    the inforequest history.
+                    """).format(admin_obj_format(obj))))
+        return warnings
+
     def delete_constraints(self, obj):
-        dependency = []
+        constraints = []
         if obj.type in [Action.TYPES.REQUEST, Action.TYPES.ADVANCED_REQUEST]:
-            dependency.append(format_html(
+            constraints.append(format_html(
                 u'{} is type {}.'.format(admin_obj_format(obj), obj.get_type_display())))
         if len(obj.branch.actions) == 1:
-            dependency.append(format_html(
+            constraints.append(format_html(
                 u'{} is the only action in the branch.'.format(admin_obj_format(obj))))
-        return dependency
+        return constraints
 
     def render_delete_form(self, request, context):
+        context[u'delete_warnings'] = self.delete_warnings(context[u'object'])
         context[u'delete_constraints'] = self.delete_constraints(context[u'object'])
         context[u'ADMIN_EXTEND_SNOOZE_BY_DAYS'] = ADMIN_EXTEND_SNOOZE_BY_DAYS
         return super(ActionAdmin, self).render_delete_form(request, context)
 
+    @decorate(short_description=u'Delete selected actions')
+    @transaction.atomic
+    def delete_selected(self, request, queryset):
+        delete_warnings = []
+        delete_constraints = []
+        outbound = InforequestEmail.objects.none()
+        inbound = InforequestEmail.objects.none()
+        for obj in queryset:
+            if obj.next_action not in queryset:
+                delete_warnings += self.delete_warnings(obj)
```

Comment:
Tato podmienka `if obj.next_action not in queryset` je trochu osemetna. Pretoze vo vseobecnosti, ked budeme robit mixin, nebude tato podmienka potrebna pre kazdy model.

Cela logika, kedy sa delete warning (alebo aj delete constraint) ma alebo nema vypisat, by mala byt vo vnutri `delete_warnings` (resp. `delete_constraints`) metody. Cize aj tato podmienka by mala byt vo vnutri `delete_warnings`. Cize `delete_warnings` malo ako parameter dostat vsetky mazane objekty, nie len jeden. Aby vedelo zobrat do uvahy aj vztahy medzi nimi.

Analogicky vies asi zovseobecnit `nested_inforequestemail_queryset`. Aby si nemusel riesit kazdy objekt zvlast a potom robit union querysetov. Ale rovno by si mohol poslat pole objektov (pripadne queryset) do `nested_inforequestemail_queryset`. Kedze vo vnutri aj tak pracujes s polom `collector.collect([obj])` v ktorom si mal doteraz iba jeden prvok. Treba ale overit, ze `collector.collect()` zafunguje spravne s polom viacerych objektov (pripadne s querysetom).

Btw, pls over, co vlastne robi `obj.next_action not in queryset`. Aby sa ti nestalo, ze to vykona SQL request na DB zakazdym pre kazdy mazany objekt.

Reply:
:ok:, 
A este co s `for obj in queryset:`? Ostane vysunuty v metode `delete_selected` alebo ho umiestnit do vnutra metody `delete_warnings` a `delete_constrints`. Jeden for cyklus viem usetrit pokial je vysunuty v metode `delete_selected`.

:ok:, collector funguje spravne aj na pole objektov, ale nie je to presne to iste. Vid. nazorny priklad, ktory som si umelo vytvoril, takato situacia by nam vsak na produkcii ani nemala vyskytnut, a asi sa snou ani netreba zaoberat. 
![20210523_161822](https://user-images.githubusercontent.com/29918752/119264385-3ec4fe80-bbe3-11eb-9a8f-f2f2f4758576.jpg)
*zmenil som inforequest zo **192** na **191** pre inforequestemail **90**
Ak by sme  v `nested_inforequestemail_queryset` pouzili `inforequest = self.get_inforequest(obj)`, tak `filter` a nasledny union by nam vratil iba inforequestemail **89**. Ak pouzijeme `inforequests = self.get_inforequest(objs)` tak nam query vrati inforequestemail **89** a tiez **90**

:ok:
Pre vypocet poctu requestov do DB som pouzil fukciu:
```
def query_count_all():
    from django.db import connections
    return sum(len(c.queries) for c in connections.all())
```
Volanim `obj.next_action` nam zvysi pocet query o jedno, `not in queryset` nema nato vplyv.

Reply:
> A este co s `for obj in queryset:`? Ostane vysunuty v metode `delete_selected` alebo ho umiestnit do vnutra metody `delete_warnings` a `delete_constrints`. Jeden for cyklus viem usetrit pokial je vysunuty v metode `delete_selected`.

IMHO praktickejsie bude ak obe metody `delete_constraints` aj `delete_warnings` dostanu rovno cele polia mazanych objektov: `def delete_warnings(self, objs)` a `def delete_constraints(self, objs)`. A bude na konkretnej implementacii pre dany model, ako to vo vnutri vyriesi.

Nerozumiem, ako chces usetrit cyklus? Mas na mysli, ze namiesto jedneho `for obj in queryset` budu dva `for obj in queryset`? To je nepodstatne. Ked bude mat Pato niekedy cas, pls pokecaj si s nim o asymptotickej zlozitosti. Ci sa nejaka vec vykona 1x alebo 2x je nepodstatne. Riesit to treba ked sa ma nejake vec vykonat Nx namiesto 1x. Ak chces iba zabespecit, aby sa `queryset` 2x neloadovalo z DB, tak si to mozes najprv explicitne loadnut do pola `objs` a potom posielat ako parameter pole namiesto querysetu. Teda za predpokladu, ze uz aj samotne Django nie je dostatocne inteligentne a vysledok `queryset` samo necachuje. To mozes overit.

Reply:
> collector funguje spravne aj na pole objektov, ale nie je to presne to iste. Vid. nazorny priklad, ktory som si umelo vytvoril, takato situacia by nam vsak na produkcii ani nemala vyskytnut, a asi sa snou ani netreba zaoberat. 
>
> Ak by sme  v `nested_inforequestemail_queryset` pouzili `inforequest = self.get_inforequest(obj)`, tak `filter` a nasledny union by nam vratil iba inforequestemail **89**. Ak pouzijeme `inforequests = self.get_inforequest(objs)` tak nam query vrati inforequestemail **89** a tiez **90**

Tu je problem v:
```
inforequestemails_qs = InforequestEmail.objects.filter(inforequest=inforequests, email__in=emails)
```
Co vytvori query (pseudokod):
```
SELECT ... FROM InforequestEmail WHERE
  inforequest=inforequest and email in (action1.email, action2.email, ...)
```
Namiesto toho by bolo treba query:
```
SELECT ... FROM InforequestEmail WHERE
  (inforequest=action1.inforequest and email=action1.email) or
  (inforequest=action2.inforequest and email=action2.email) or ...
```
Aby sa zabespecilo, ze sa selectnu iba `InforequestEmail` patriace mazanym akciam. Ze ak niektory email ma viacero `InforequestEmail` priradenych viacerym inforequestom, tak sa zmaze iba jedna instancia `InforequestEmail` patriaca mazanej akcii.

Reply:
>  Pre vypocet poctu requestov do DB som pouzil fukciu:

Pre sledovanie ake DB queries sa vykonaju sa zide v `settings/server_local.py` odkomentovat `django.db.backends` logger a jeho `file_db` handler. A potom pomocou `tail -f logs/db.log` pozerat, ake query DB vykonava. A ci sa v ramci jedneho API requestu nejaka query neopakuje moc casto.

Reply:
:ok: aj novy queryset vykona rovnaky pocet query do databazy ako ten predosly.

Reply:
@viliambalaz Este sme nedoriesili toto z https://github.com/martinmacko47/chcemvediet/pull/382#discussion_r638058490 o par commentov vyssie v tomto threade:

> IMHO praktickejsie bude ak obe metody `delete_constraints` aj `delete_warnings` dostanu rovno cele polia mazanych objektov: `def delete_warnings(self, objs)` a `def delete_constraints(self, objs)`. A bude na konkretnej implementacii pre dany model, ako to vo vnutri vyriesi.